test/integration_test.rb

```diff
@@ -41,36 +41,81 @@ def test_site_validation
-      assert_equal 302, response.status
-      assert_match /\A#{Regexp.quote("/auth/failure?message=invalid_site")}/, response.location
+      response = callback(sign_params(shop: shop, code: code))
+      assert_auth_failure(response, 'invalid_site')
     end
   end
 
   def test_callback
     access_token = SecureRandom.hex(16)
     code = SecureRandom.hex(16)
+    expect_access_token_request(access_token)
+
+    response = callback(sign_params(shop: 'snowdevil.myshopify.com', code: code))
+
+    assert_callback_success(response, access_token, code)
+  end
+
+  def test_callback_with_legacy_signature
+    access_token = SecureRandom.hex(16)
+    code = SecureRandom.hex(16)
+    expect_access_token_request(access_token)
+
+    response = callback(sign_params(shop: 'snowdevil.myshopify.com', code: code).merge(signature: 'ignored'))
+
+    assert_callback_success(response, access_token, code)
+  end
+
+  def test_callback_custom_params
+    access_token = SecureRandom.hex(16)
+    code = SecureRandom.hex(16)
     FakeWeb.register_uri(:post, "https://snowdevil.myshopify.com/admin/oauth/access_token",
                          body: JSON.dump(access_token: access_token),
                          content_type: 'application/json')
 
-    response = callback(shop: 'snowdevil.myshopify.com', code: code)
+    now = Time.now.to_i
+    params = { shop: 'snowdevil.myshopify.com', code: code, timestamp: now, next: '/products?page=2&q=red%20shirt' }
+    encoded_params = "code=#{code}&next=/products?page=2%26q=red%2520shirt&shop=snowdevil.myshopify.com&timestamp=#{now}"
+    params[:hmac] = OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA256.new, @secret, encoded_params)
 
-    token_request_params = Rack::Utils.parse_query(FakeWeb.last_request.body)
-    assert_equal token_request_params['client_id'], '123'
-    assert_equal token_request_params['client_secret'], '53cr3tz'
-    assert_equal token_request_params['code'], code
+    response = callback(params)
 
-    assert_equal 'snowdevil.myshopify.com', @omniauth_result.uid
-    assert_equal access_token, @omniauth_result.credentials.token
-    assert_equal false, @omniauth_result.credentials.expires
+    assert_callback_success(response, access_token, code)
+  end
 
-    assert_equal 200, response.status
-    assert_equal "OK", response.body
+  def test_callback_rejects_invalid_hmac
+    @secret = 'wrong_secret'
+    response = callback(sign_params(shop: 'snowdevil.myshopify.com', code: SecureRandom.hex(16)))
+
+    assert_auth_failure(response, 'invalid_signature')
+  end
+
+  def test_callback_rejects_old_timestamps
+    response = callback(sign_params(shop: 'snowdevil.myshopify.com', code: SecureRandom.hex(16), timestamp: Time.now.to_i - 5 * 60 - 1))
+
+    assert_auth_failure(response, 'invalid_signature')
+  end
+
+  def test_callback_rejects_missing_hmac
+    code = SecureRandom.hex(16)
+
+    response = callback(shop: 'snowdevil.myshopify.com', code: code, timestamp: Time.now.to_i - 5 * 60 - 1)
```

Comment:
Delete `- 5 * 60 - 1`; otherwise this test is hitting the timestamp check, not the hmac validation.